### PR TITLE
CDW Round 47

### PR DESCRIPTION
**Pull request by cmperro**
This PR is to complete the work needed to enable (the happy path for) CDW creation and teardown.

At a high level it allows for:

- Creation of a CDW environment
- Creation of a default CDW database catalog
- Creation of non-default CDW database catalogs
- Loading demo data in non-default CDW database catalogs
- Creation of Hive and Impala VWs
- Specifying VW autoscaling min and max groups
- Specifying VW template/size
- Specifying VW tags
- Specifying Common and Application configs for VWs
- Enabling SSO for VWs
- Applying Group RBAC for VWs

It requires [CDPY PR 44](https://github.com/cloudera-labs/cdpy/pull/44)

### CODE REVIEW

**Review by wmudge — CHANGES_REQUESTED** [review]
Conflicts with this PR -- I'm guessing a rebase is in order?

**Comment by cmperro** [comment]
I did one when I started this journey (~1.5 weeks ago). I’m guessing there’s been merged into devel since then?

**Comment by cmperro** [comment]
Rebase... done.

CDW CLI... back to working.

Please test!

**Review by Chaffelson — CHANGES_REQUESTED** [review]

File: roles/runtime/tasks/setup_aws.yml [hunk 1]
```diff
@@ -35,31 +35,107 @@
...
-      register: __dw_builds
+      register: __dw_vw_builds
+      loop: "{{ run__dw_vw_configs }}"
+      loop_control:
+        pause: 30 #Avoid dual compactors
+        loop_var: __dw_vw_config
+        label: "{{ __dw_vw_config.name }}"
 
-    - name: Wait for CDP DW cluster setup to complete
+    - name: Wait for CDP DW Virtual warehouses setup to complete
```
[the remark this answers is not in the document]
Comment (cmperro) [reply roles/runtime/tasks/setup_aws.yml:131]: I hit a variant of this bug, it's why the task above that creates the VWs has a 30 second "pause".

File: roles/runtime/tasks/setup_aws.yml [hunk 2]
```diff
@@ -35,31 +35,107 @@
         label: "{{ __aws_instance_item.instance_id }}"
       loop: "{{ __aws_instance_info.instances }}"
 
-- name: Setup CDP DW cluster on AWS
+- name: Setup CDP Data Warehouse on AWS
   when: run__include_dw
   tags: dw
   block:
-    - name: Execute CDP DW cluster setup
+    - name: Execute CDP Data Warehouse environment setup
```
[the remark this answers is not in the document]
Comment (cmperro) [reply roles/runtime/tasks/setup_aws.yml:42]: I debated here... the DW terminology is to overload the term "environment": https://docs.cloudera.com/data-warehouse/cloud/aws-environments/topics/dw-activating-environments-4-data-catalogs.html

File: docs/configuration.yml [hunk 3]
```diff
@@ -230,7 +259,7 @@ globals:
   dynamic_inventory:
     vm:
       count:
-      os:
```
[the remark this answers is not in the document]
Comment (cmperro) [reply docs/configuration.yml:233]: I don't even know why that change would have snuck in... I'll update it.

File: roles/common/defaults/main.yml [hunk 4]
```diff
@@ -111,11 +111,17 @@ common__aws_public_subnet_ids:            "{{ infra.aws.vpc.existing.public_subn
...
 common__aws_region:                       "{{ infra.aws.region | default('eu-west-1') }}"
 common__aws_role_suffix:                  "{{ infra.aws.role.suffix | default(common__role_suffix) }}"
+
 common__aws_datalake_admin_role_name:     "{{ env.aws.role.name.datalake_admin | default([common__namespace, common__aws_datalake_admin_suffix, common__aws_role_suffix] | join('-')) }}"
 common__aws_datalake_admin_suffix:        "{{ env.aws.role.label.datalake_admin | default(common__datalake_admin_suffix) }}"
 common__aws_idbroker_role_name:           "{{ env.aws.role.name.idbroker | default([common__namespace, common__aws_idbroker_suffix, common__aws_role_suffix] | join('-')) }}"
 common__aws_idbroker_suffix:              "{{ env.aws.role.label.idbroker | default(common__idbroker_suffix) }}"
 
+
+common__aws_vpc_id:                       "{{ infra.aws.vpc.existing.vpc_id | default('') }}"
```
[the remark this answers is not in the document]
Comment (cmperro) [reply roles/common/defaults/main.yml:121]: Got it.

File: roles/runtime/tasks/initialize_teardown.yml [hunk 5]
```diff
@@ -24,11 +24,37 @@
...
+    - run__include_dw
+    - not run__force_teardown | bool or not run__dw_force_delete | bool
+  block:
+    - name: Discover CDP Data Warehouse cluster
+      cloudera.cloud.dw_cluster_info:
+        env: "{{ run__env_name }}"
+      register: __dw_list
+
+    - name: Initialize CDP Data Warehouse cluster id
+      ansible.builtin.set_fact:
```
[the remark this answers is not in the document]
Comment (cmperro) [reply roles/runtime/tasks/initialize_teardown.yml:44]: Bumping this out of the block. The entire block should be skipped if there's not DW cluster.

File: roles/runtime/tasks/setup_aws.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (Chaffelson) [reply roles/runtime/tasks/setup_aws.yml:131]: Would fixing the setup check to also include 'Stopped' as a valid deployment completion state fix your bug as well, and thus we could remove the pause?

File: roles/runtime/tasks/setup_aws.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (Chaffelson) [reply roles/runtime/tasks/setup_aws.yml:42]: I suggest we clarify by calling it 'Activate Environment for Cloudera Data Warehouse' then, if they insist on not referencing the K8s setup.

**Comment by cmperro** [comment]
Comments addressed and rebase done.

**Review by Chaffelson — APPROVED** [review]

**Review by wmudge — APPROVED** [review]

**Pull request merged**